### PR TITLE
opencv 4.x: refactor to allow fine-grained selection of each module/contrib

### PR DESCRIPTION
**Pull request by SpaceIm**
- Refactoring to allow disabling each module independently:

  - Add options with consistent names to allow enabling/disabling:
 
    - main modules (except `core` obviously), not only `gapi` & `dnn`: `calib3d`, `features2d`, `flann`, `highgui`, `imgproc`, `imgcodecs`, `ml`, `objdetect`, `photo`, `stitching`, `video` and `videoio`.
      As a consequence `with_ade` option is deprecated in favor of `gapi` option.

    - each extra module (~55 libraries).
      As a consequence `contrib` option is deprecated. `contrib_sfm` & `contrib_freetype` options are also deprecated in favor of `sfm` & `freetype` options respectively (for option naming consistency).

  - Due to the huge number of optional modules and internal dependencies between them, it may be very hard for consumers to find the good combination allowing to build a specific module (raising for each bad tuples of options would be a nightmare for users). Therefore when a module is enabled by users, opencv modules of its internal dependency graph are also enabled if they were disabled, and recipe displays a warning for each module enabled with this mechanism (if they were disabled by default or by consumers).

  - Several issues had to be fixed to allow this refactoring:

    - Properly declare in `package_info()` these missing contrib modules (they were built when contrib option was True, but not properly declared as components):
      - `dnn_objdetect` & `text`
      - `dnn_superres` since 4.1.2
      - `mcc` since 4.5.0
      - `wechat_qrcode` since 4.5.2
      - `barcode` since 4.5.3

    - Allow to build 3 more extra modules `cvv`, `hdf` & `ovis`.

    - Allow to build `objdetect` main module without `dnn` module in 4.5.5 (it can't be built upstream without `dnn` only in 4.5.4 & 4.5.5). It has been "fixed" upstream in OpenCV 4.6.0 (see https://github.com/opencv/opencv/pull/21369).
  It fixes few things in this recipe:
      - a regression introduced by https://github.com/conan-io/conan-center-index/pull/11009 (an incorrect fix for https://github.com/conan-io/conan-center-index/issues/10997). Indeed, `objdetect` can be built without dnn in all OpenCV versions before 4.5.4, so this PR was breaking these versions when `dnn` option was False by not exposing `opencv_objdetect` and few contrib components.
      - avoid a bug (which is also a regression coming from PR linked above) where quirc is in requirements, but not required by any component in `package_indo()` when `dnn=False` (the one https://github.com/conan-io/conan-center-index/pull/13215 was trying to address, but badly due to previous regression).

    - Fix "requires" of `photo` & `stitching` components when cuda is enabled.

  - Refactor tests to take into account these changes, it's more modular. It allows to test each component in isolation and more easily detect issues in opencv recipe.

- Few options have been added in the meantime:

  - `world` which aggregates all modules (except `img_hash` & `sfm`) in one lib (aggregated modules still have a dedicated imported target to properly mimic official OpenCV config file, but these targets depend on `opencv_world`).

  - `with_tesseract` (enabled by default): build `text` extra module with `tesseract` support. Option is removed if `text` module is not built.

  - `with_gdal` & `with_gdcm` (disabled by default): optional dependencies of `imgcodecs` main module. Options are removed if `imgcodecs` module is not built.

  - `with_qt` (disabled by default): optional dependency of `highgui` main module, and mandatory for `cvv` extra module. Option is removed if `highgui` and `cvv` module are not built.

  - `with_vulkan` (disabled by default): optional dependency of `dnn` main module. Option is removed if `dnn` module is not built.

  - `with_opencl` (disabled by default): optional global dependency.

- Few compat improvements:

  - compat with protobuf >= 3.18 in OpenCV < 4.5.4 (backport of https://github.com/opencv/opencv/pull/20386)
  - compat with OpenEXR 3.x in OpenCV < 4.5.3

- conan v2 fix:

  - use `self.dependencies["gtk"].options.version` instead of `self.options["gtk"].version`
  - add `res` dir to `resdirs` attribute of components in `package_info()`

closes https://github.com/conan-io/conan-center-index/issues/4265
closes https://github.com/conan-io/conan-center-index/issues/7439
closes https://github.com/conan-io/conan-center-index/issues/9127
closes https://github.com/conan-io/conan-center-index/issues/14816
closes https://github.com/conan-io/conan-center-index/issues/15581
closes https://github.com/conan-io/conan-center-index/issues/16726
closes https://github.com/conan-io/conan-center-index/issues/17042
closes https://github.com/conan-io/conan-center-index/issues/17182
closes https://github.com/conan-io/conan-center-index/issues/17368

---

- [ ] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [ ] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by Talkless** [comment]
@SpaceIm Big thanks for working on this!

**Comment by SpaceIm** [comment]
@uilianries @prince-chrismc @jwillikers @SSE4 Could you review this PR please? I know it's huge, but I would like to merge these fixes before packaging 4.7.0

**Comment by SSE4** [comment]
@SpaceIm the PR is indeed huge.
more, it contains many custom patches, that makes it even harder to review.
do you think moving some logic about components and dependencies to declarative json/yaml file would help to simplify the code?

**Comment by SpaceIm** [comment]
> Do you think moving some logic about components and dependencies to declarative json/yaml file would help to simplify the code?

It would mean a yml file of 1000 lines per version.

**Review by jwillikers — DISMISSED** [review]

**Review by jwillikers — APPROVED** [review]

**Review by samuel-emrys — APPROVED** [review]

**Comment by yekmen** [comment]
Just for information that "world" option isn't work. 
You forgot to set "BUILD_opencv_world"

**Comment by SpaceIm** [comment]
Indeed, I've lost this option in https://github.com/conan-io/conan-center-index/pull/15193/commits/a74ac565e880bfcd2a407ab6dde16279cf2ad6b3, I'll fix that.

**Comment by SpaceIm** [comment]
```
12/14 Test #14: test_videoio .....................***Failed    0.01 sec
/home/conan/w/prod/BuildSingleReference/cci-335c42c4/recipes/opencv/4.x/test_v1_package/build/fc926f3ef0846225112e32da37080ac8b54a6e8c/bin/test_videoio: error while loading shared libraries: libva-drm.so.1: cannot open shared object file: No such file or directory
```

WTF

File: recipes/opencv/4.x/conanfile.py [hunk 1]
```diff
@@ -148,101 +288,848 @@ def config_options(self):
...
+            },
+            "cudabgsegm": {
+                "is_built": self.options.cudabgsegm,
+                "mandatory_options": ["with_cuda", "video"],
+                "requires": ["opencv_video"] + ipp(),
+            },
+            "cudacodec": {
+                "is_built": self.options.cudacodec,
+                "mandatory_options": ["with_cuda", "videoio"],
+                "requires": ["opencv_core", "opencv_videio", "opencv_cudev"] + ipp(),
```
Comment (tankeco) [inline recipes/opencv/4.x/conanfile.py:572]: typo here

Comment (SpaceIm) [reply recipes/opencv/4.x/conanfile.py:572]: thanks

File: recipes/opencv/4.x/conandata.yml [hunk 2]
```diff
@@ -60,6 +94,25 @@ patches:
...
+    - patch_file: "patches/4.1.2-0009-sfm-deps.patch"
+      patch_description: "Fix link to dependencies of sfm module"
+      patch_type: "conan"
+    - patch_file: "patches/4.1.2-0010-compat-protobuf-ge-3.18.patch"
+      patch_description: "Compatibility with protobuf >= 3.18"
+      patch_type: "portability"
+      patch_source: "https://github.com/opencv/opencv/pull/20386"
+    - patch_file: "patches/4.1.2-0011-compat-openexr3.patch"
+      patch_description: "Compatibility with OpenEXR 3.x"
+      patch_type: "portability"
```
Comment (uilianries) [inline recipes/opencv/4.x/conandata.yml:118]:
```suggestion
      patch_type: "portability"
      patch_source: "https://github.com/opencv/opencv/pull/19970"
```

**Review by uilianries — COMMENTED** [review]
Looking good, just an info about a patch

**Review by uilianries — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**